demos/howtos/myula.py

```diff
@@ -0,0 +1,217 @@
+# %%
+# Then we can define the likelihood with
+likelihood=y(y=y_obs)
+
+# %%
+# Implicit prior definition
+# ---------------------------------
+# To apply MYULA, we need to define the implicit prior :math:`\pi_{\texttt{stength_smooth}}(x)`.
+# Evaluating this surrogate prior is doable but too intensive from
+# a computational point of view as it requires to solve an optimization problem.
+# However to apply MYULA, we only require access to
+# :math:`\operatorname{prox}_{\texttt{strength_reg}\ TV}^{\texttt{strength_smooth}}`.
+#
+# As suggested by Durmus et al. (https://arxiv.org/pdf/1612.07471), we set the
+# smoothing parameter :math:`\texttt{strength_smooth} \approx \texttt{sigma2}`,
+# ie :math:`\texttt{strength_smooth}= 0.5 \ \texttt{sigma2}`.
+#
+# We set the regularization parameter to :math:`\texttt{stength_reg}=10`.
+
+# %%
+# Regularization and smoothing parameters definition
+strength_reg=10
+strength_smooth=0.5*sigma2
+# %%
+# To estimate :math:`\operatorname{prox}_{\texttt{strength_reg}\  TV}^{\texttt{strength_smooth}}`
+# we use the implementation provided by Scikit-Image. But we can use any solver
+# to compute this quantity.
+#
+# We emphasize that we have for any :math:`g`
+#
+# .. math::
+#       \operatorname{prox}_{\texttt{strength_reg}\  g}^{\texttt{strength_smooth}} = \operatorname{prox}_{g}^{\texttt{weight}} ,
+#
+# with :math:`\texttt{weight} = \texttt{strength_reg} \times  \texttt{strength_smooth}`.
+from skimage.restoration import denoise_tv_chambolle
+def prox_g(x, strength_reg=None, strength_smooth=None):
+    weight=strength_reg*strength_smooth
+    return denoise_tv_chambolle(x, weight=weight, max_num_iter=100), None
+# %%
+# We save all the important variables into the variable
+# :math:`\texttt{denoiser_kwargs}`.
+denoiser_kwargs={}
+denoiser_kwargs["strength_reg"]=strength_reg
+denoiser_kwargs["strength_smooth"]=strength_smooth
+# %%
+# Now we can define our implicit prior.
+denoise_regularizer = DenoiseRegularizer(
+    prox_g,
+    strength_smooth=strength_smooth,
+    denoiser_kwargs=denoiser_kwargs,
+    geometry=likelihood.model.domain_geometry
+)
+
+# %%
+# Implicitly defined posterior distribution
+# -----------------------------------------
+# We can now define the implicitly defined posterior distribution as follows:
+posterior=Posterior(likelihood, denoise_regularizer)
+
+# %%
+# Parameters of the MYULA sampler
+# ------------------------------
+# We let run MYULA for :math:`\texttt{Ns}=10^4`
+# iterations. We discard the :math:`\texttt{Nb}=1000` first burn-in samples of
+# the Markov chain. Furthermore, as MCMC methods generate
+# correlated samples, we also perform a thinning: we only consider 1 samples
+# every :math:`\texttt{Nt}=20`
+# samples to compute our quantities of interest.
+# :math:`\texttt{scale}` is set wrt the recommendation of Durmus et al.
+# (https://arxiv.org/pdf/1612.07471).
+Ns=10000
+Nb=1000
+Nt=20
+# Step-size of MYULA
+scale=0.9/(1/sigma2 + 1/strength_smooth)
+# %%
```

Comment:
Also, if this is a sensible default value for the scale, could it be set as a default inside the sampler?

Reply:
It is a good point. We need to talk about it with Amal to see how to proceed.
However, I just want to point out that in some cases, estimating the Lipschitz constant of the gradient of the log-posterior density is not always doable. For example, we cannot estimate the Lipschitz constant for the NLMeans denoiser.

Reply:
Status update: now in the MYULA, PnPULA samplers, the default value of scale is consistent with ULA (same value)